### PR TITLE
Report allocation errors as panics

### LINKED ISSUE #51245: Tracking issue for the OOM hook
PR #50880 added an API to override the std OOM handler, similarly to the panic hook. This was discussed previously in issue #49668, after PR #50144 moved OOM handling out of the `Alloc`/`GlobalAlloc` traits. The API is somewhat similar to what existed before PR #42727 removed it without an explanation. This issue tracks the stabilization of this API.

Defined in the `std::alloc` module:

<del>

```rust
pub fn set_oom_hook(hook: fn(Layout) -> !);
pub fn take_oom_hook() -> fn(Layout) -> !;
```

</del>

```rust
pub fn set_alloc_error_hook(hook: fn(Layout));
pub fn take_alloc_error_hook() -> fn(Layout);
```

CC @rust-lang/libs, @SimonSapin 

## Unresolved questions
- [x] ~~Name of the functions. The API before #42727 used `_handler`, I made it `_hook` in #50880 because that's the terminology used for the panic hook (OTOH, the panic hook returns, contrary to the OOM hook).~~  #51264
- [ ] Should this move to its own module, or stay in `std::alloc`?
- [x] Interaction with unwinding. `alloc::alloc::oom` is marked `#[rustc_allocator_nounwind]`, so theoretically, the hook shouldn't panic (except when panic=abort). Yet if the hook does panic, unwinding seems to happen properly except it doesn't.

### LINKED ISSUE #51540: Tracking issue for the #[alloc_error_handler] attribute (for no_std + liballoc)
This attribute is mandatory when using the `alloc` crate without the `std` crate. It is used like this:

```rust
#[alloc_error_handler]
fn foo(_: core::alloc::Layout) -> ! {
    // …
}
```

Implementation PR: https://github.com/rust-lang/rust/pull/52191

## Blocking issues

* [ ] #83430

## Original issue:

----

In a `no_std` program or staticlib, linking to the `alloc` crate may cause this error:

```
error: language item required, but not found: `oom`
```

This is fixed by providing the `oom` lang item, which is is normally provided by the `std` crate (where it calls a dynamically-settable hook https://github.com/rust-lang/rust/issues/51245, then aborts). This is called by `alloc::alloc::handle_alloc_error` (which is called by `Vec` and others on memory allocation failure).

```rust
#![feature(lang_items)]

#[lang = "oom"]
extern fn foo(_: core::alloc::Layout) -> ! {
    // example implementation based on libc
    extern "C" { fn abort() -> !; }
    unsafe { abort() }
}
```

However, defining a lang item is an unstable feature.

Possible solutions include:

1. Add and stabilize a dedicated attribute similar to [the `#[panic_implementation]` attribute](https://rust-lang.github.io/rfcs/2070-panic-implementation.html#panic_implementation):

  ```rust
  #[alloc_error_handler]
  fn foo(_: core::alloc::Layout) -> ! {
      // …
  }
  ```

  The downside is that this is one more mandatory hoop to jump through for `no_std` program that would have been fine with a default hook that aborts.

2. ~~Move `std`’s dynamically-settable hook into `alloc`: https://github.com/rust-lang/rust/pull/51607. The downside is some mandatory space overhead.~~

### PR DESCRIPTION
**Pull request by Amanieu**
OOM is now reported as a panic but with a custom payload type (`AllocErrorPanicPayload`) which holds the layout that was passed to `handle_alloc_error`.

This should be review one commit at a time:
- The first commit adds `AllocErrorPanicPayload` and changes allocation errors to always be reported as panics.
- The second commit removes `#[alloc_error_handler]` and the `alloc_error_hook` API.

ACP: https://github.com/rust-lang/libs-team/issues/192

Closes #51540
Closes #51245

### CODE REVIEW

**Review by Sheylas1 — APPROVED** [review]

**Review by lnicola — APPROVED** [review]

File: library/alloc/src/alloc.rs [hunk 1]
```diff
@@ -349,28 +351,82 @@ pub(crate) const unsafe fn box_free<T: ?Sized, A: ~const Allocator + ~const Dest
...
+
+    // Hack to work around issues with the lifetime of Arguments.
+    match format_args!("memory allocation of {} bytes failed", layout.size()) {
+        fmt => {
+            // Create a PanicInfo with a custom payload for the panic handler.
+            let payload = AllocErrorPanicPayload { layout };
+            let mut pi = core::panic::PanicInfo::internal_constructor(
+                Some(&fmt),
+                core::panic::Location::caller(),
+                unsafe { __rust_alloc_error_handler_should_panic != 0 },
```
Comment (RalfJung) [inline library/alloc/src/alloc.rs:415]: This is a bit obscure due to the unclear meaning of this boolean.
```suggestion
                /* can_unwind */ unsafe { __rust_alloc_error_handler_should_panic != 0 },
```
Or a `let can_unwind = ...` earlier.

File: compiler/rustc_feature/src/removed.rs [hunk 2]
```diff
@@ -47,6 +47,8 @@ declare_features! (
 
     (removed, advanced_slice_patterns, "1.0.0", Some(62254), None,
      Some("merged into `#![feature(slice_patterns)]`")),
+     /// Allows defining an `#[alloc_error_handler]`.
```
Comment (bjorn3) [inline compiler/rustc_feature/src/removed.rs:50]:
```suggestion
    /// Allows defining an `#[alloc_error_handler]`.
```
nit

**Review by davidtwco — APPROVED** [review]
LGTM, r=me if you're just waiting on reviewer

**Comment by Amanieu** [comment]
FCP is complete, this is now ready to merge.

**Comment by Manishearth** [comment]
Could we please decouple the removal of the attribute from the actual switchover?


This PR causes a hard edge on nightly for embedded development: code before this PR _must_ have `#[alloc_error_handler]`, and code after this PR is forbidden from having `#[alloc_error_handler]`.

I know this is an unstable API, but can we go through a deprecation period? This kind of hard edge is a major pain to deal with.


Using future incompat for this would be ideal.

**Comment by Manishearth** [comment]
I've opened https://github.com/rust-lang/rust/pull/110733

**Pull request merged**